### ISSUE TITLE
[Perf] Regressions in Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable

### ISSUE THREAD

**Issue by DrewScoggins** [dotnet/runtime#54170]
### Run Information

Architecture | arm64
-- | --
OS | Windows 10.0.19041
Baseline | [5f15498d937653986821d636c04cd03c58d4bb4c](https://github.com/dotnet/runtime/commit/5f15498d937653986821d636c04cd03c58d4bb4c)
Compare | [293d47286161827ecc0d6e96f9eb7b82dafe8307](https://github.com/dotnet/runtime/commit/293d47286161827ecc0d6e96f9eb7b82dafe8307)
Diff | [Diff](https://github.com/dotnet/runtime/compare/5f15498d937653986821d636c04cd03c58d4bb4c...293d47286161827ecc0d6e96f9eb7b82dafe8307)

### Regressions in Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[Scoped - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable.Scoped.html>) | 84.69 ns | 107.57 ns | 1.27 | 0.34 | | |
[Transient - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable.Transient.html>) | 160.48 ns | 7.33 μs | 45.66 | 0.41 | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/06_02_2021/refs/heads/main_arm64_Windows%2010.0.19041_Regression/Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable_1.png>)
![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/06_02_2021/refs/heads/main_arm64_Windows%2010.0.19041_Regression/Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable_2.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/06_02_2021/refs/heads/main_arm64_Windows%2010.0.19041_Regression/Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-8e06945c-cd16-4c3b-b4cc-c03c7c9228764f32958b30c40c486/8a460bcc-a106-4a24-96a6-3360f78b1816.zip?sv=2019-07-07&se=2021-06-25T01%3A37%3A08Z&sr=c&sp=rl&sig=FNG3xf5TGQzLF2Lvg9L8BAiDqyBRSfnIDFXtbrurv1g%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-06ac4bbf-d1b0-452b-84d1-752c2f8387facd29e5e30f04a16b0/7483e558-831a-41a5-95c4-55edf1c70607.zip?sv=2019-07-07&se=2021-06-25T11%3A12%3A29Z&sr=c&sp=rl&sig=lMczQtPcK70tSbKDYIluSXHN%2FEqXVhVsaAR0E9dGLDc%3D>)

#### Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable.Scoped

```log

```
#### Microsoft.Extensions.DependencyInjection.GetServiceIEnumerable.Transient

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

**Comment by DrewScoggins** [comment]
https://github.com/dotnet/runtime/pull/52140

**Comment by eerhardt** [comment]
@davidfowl @pakrym

**Comment by maryamariyan** [comment]
here are the latest charts:

![newplot (9)](https://user-images.githubusercontent.com/5897654/128100267-d2051607-629a-4557-84ac-47b8264c67c4.png)

![newplot (10)](https://user-images.githubusercontent.com/5897654/128100303-c6687e9d-d31c-4b4a-9267-6883b4062978.png)

**Comment by davidfowl** [comment]
So only enumerables regressed?

**Comment by pakrym** [comment]
Did my change prevent IEnumerable accessors from being compiled? The jump is very high for transient.

**Comment by maryamariyan** [comment]
yes the perf jump lists these PRs https://statics.teams.cdn.office.net/evergreen-assets/safelinks/1/atp-safelinks.html, from which #52140 is listed in.

**Comment by davidfowl** [comment]
We need to look into it, last time we regressed the perf it was throwing exceptions and nobody knew 😄

**Comment by pakrym** [comment]
Yeah, I think it might be IL building process failing. we need to add a debug.Fail somewhere :)

**Comment by pakrym** [comment]
We do have it in https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/DynamicServiceProviderEngine.cs#L42

**Comment by davidfowl** [comment]
@pakrym spot the bug https://github.com/dotnet/runtime/blob/7c8f5fbee22030e69ca69615a157597cc035e854/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs#L170

**Comment by eerhardt** [comment]
@maryamariyan is working on getting perf numbers with this fixed. My hope is it fixes these regressions.

**Comment by pakrym** [comment]
Oh no, 🤦 `ImplementationType`..

**Comment by maryamariyan** [comment]
> @maryamariyan is working on getting perf numbers with this fixed. My hope is it fixes these regressions.

yes, more info available in this comment https://github.com/dotnet/runtime/issues/56781#issuecomment-893677784

**Comment by eerhardt** [comment]
Fixed by #57155.
